Enhance Configuration & Housekeeping

thecacho opened the pull request:
# Enhance Configuration & Housekeeping
Hi, before going into the changes i want to say that i never touched Lua before... i did theses changes trying to learn lua from scratch, while improving my new setup of Neovim (LazyNvim), using this plugin. And i have to say, this is an awesome package, and a MUST for any nvim developer. 🙌🏼 

### Features
The main changes were:
1.  The addition of `Type Definitions & enums`, in which the `ConfigSchema` (& local functions) are described and documented... This enhances the user setup, by having some auto-completion, i updated the `LazyVim` example to demonstrate the "implementation" of.

2. While i was implementing the `Types` constraints i renamed some ambiguous variables, like `config` -> `default_config`, and other minor changes (Gral. Cleanup)...

3. Add missing configurable options to the Readme.md

thecacho committed:
feat(types): add types & gral cleanup

This commit enforces the arguments, variables and functions types/requirements & naming

thecacho committed:
docs(readme): update options example

Add missing options

thecacho committed:
docs(readme): add lazynvim lua type completion

thecacho committed:
style: follow variable naming convention

- Changed from camel case to _ separation

thecacho committed:
refactor(config): protect & rename the default_config

thecacho committed:
refactor(config): move default config to config

To fix the default config test, the local variable default_config, was moved into config. Making the instance of that private in the main carbon-now.lua but still accessible from the carbon_now_spec.lua

thecacho commented on README.md:
```diff
@@ -41,18 +47,19 @@ The plugin comes with the following default configs:
+    line_height = "133%",
+    line_numbers = true,
+    theme = "monokai",
+    titlebar = "Made with carbon-now.nvim",
```
The `titlebar` is missing on the realmente.md configuration 👀

thecacho reviewed (commented):
Found a typo (misspell) on 
`Couldn't found` -> `Couldn't find`

thecacho committed:
fix: typo on error message

From `Couldn't found` -> `Couldn't find`

ellisonleao commented on lua/carbon-now.lua:
```diff
@@ -1,28 +1,11 @@
 local carbon = {}
 
--- default configs
-carbon.config = {
```
why create another file for this?

thecacho replied:
Hi @ellisonleao, i came as fast as possible, from my work, so. I saw this pattern a lot, in other lua scripts/plugins, like `nvim-cmp` or `gitsigns.nvim`, where the default configuration is abstracted into it's own file...  having a separated `default.lua` file to define the default configuration for your Neovim plugin is a good practice. It promotes code organization, reusability, extensibility, and allows future users to customize the plugin to their needs on demand.

This way, you can require only the configuration, without any bias to mutate the result or run any extra code along the way...

So let's do a brief (long) Recap of what i think:
1. Separation of Concerns: By abstracting the default configuration into a separate file, you're keeping the configuration logic isolated from the rest of the plugin code. This promotes a modular & organized structure making it easier to maintain and understand.

2. Reusability: Having the default configuration in its own file makes it reusable across different parts of your plugin.If there are multiple components or modules that require access to the default configuration, they can simply import the default.lua file and use the default_config table. (Actually, that was the main reason here, when i include this default settings on your tests)

3. Extensibility: Separating the default configuration allows for easy extensibility in the future. If you decide to introduce additional configuration options or modify existing ones, you can do so by updating the default.lua file without having to modify other parts of your plugin code, creating different flavors of it,

ellisonleao replied:
hey @Domenic-MZS those are good and valid points, but this is a very small plugin and i don't see the need for it. Can you please just leave as it is now please?

thecacho replied:
Okay, will fix it in the next commit

thecacho replied:
Fixed with [8aa5de1](https://github.com/ellisonleao/carbon-now.nvim/pull/4/commits/8aa5de1d050c41f69bdedd20aebaff80a4f43313)

ellisonleao commented on lua/carbon-now.lua:
```diff
@@ -33,29 +16,34 @@ local function query_param_encode(str)
-    end
-    table.insert(params, k .. "=" .. query_param_encode(v))
+    local parsed_value = tostring(v)
+    local encoded_value = query_param_encode(parsed_value)
```
since we are not having any errors or error treatment here, i believe this var is unnecessary, can you just put the calculation back in the `table.insert` call please?

thecacho replied:
You're right, i will revert it on the next commit

thecacho replied:
Fixed on [8aa5de1](https://github.com/ellisonleao/carbon-now.nvim/pull/4/commits/8aa5de1d050c41f69bdedd20aebaff80a4f43313)

ellisonleao commented on lua/types/init.lua:
```diff
@@ -0,0 +1,5 @@
+local types = {}
+
+types.cn = require("types.cn")
```
is `types.cn` module really necessary or can you just drop the code here?

thecacho replied:
It is not strictly necessary in a manner of speaking...  Before i made this Pull Request, i checked for similar requests, and i found this along the way #1 . Which suggested other sources/modules. 

Therefore, i considered that there's a possibility in the future that the codebase could grow (expand modules or be reused separately), and it's better to keep them separate for better organization and maintainability with the different modules types like: `carbon-now`, `ray.so` or `bin-share`.

However, it makes more sense to keep it simple in one file for the time being... i could revert that if you want it, so LMK (let me know).

thecacho replied:
> This one I do admit is my fault 😅 a little. I was too excited while doing it.

ellisonleao replied:
> However, it makes more sense to keep it simple in one file for the time being... i could revert that if you want it, so LMK (let me know).

Yeah, just drop that module code over there and it should be fine. No need for that extra module

thecacho replied:
Fixed on [8aa5de1](https://github.com/ellisonleao/carbon-now.nvim/pull/4/commits/8aa5de1d050c41f69bdedd20aebaff80a4f43313)

ellisonleao commented on lua/carbon-now.lua:
```diff
@@ -91,37 +80,46 @@ local function get_open_command()
     return "start"
   end
+
+  error("Couldn't find a launch command")
```
```suggestion
  vim.api.nvim_err_writeln("[carbon-now.nvim] Couldn't find a launch command")
```

thecacho replied:
Just googled it, you're right, good catch. I didn't know the difference between them. I really want to thank you for your time and feedbacks, for real yo' ! 🏆 💪🏼  

and yes i didn't forgot it, i will fix it on the next commits 🙌🏼

thecacho replied:
Fixed on [8aa5de1](https://github.com/ellisonleao/carbon-now.nvim/pull/4/commits/8aa5de1d050c41f69bdedd20aebaff80a4f43313)

thecacho committed:
fix: apply review feedback and fixes

- replace the error method with vim.api.nvim_err_writeln (prevent fatal err)
 - revert default config abstraction
 - revert encode_parameters variables -> calculations occur in the `table.insert` call
 - merge types.cn with types

reviewer: @ellisonleao

ellisonleao commented on lua/carbon-now.lua:
```diff
@@ -1,7 +1,9 @@
 
--- default configs
+-- default config
+---@type cn.ConfigSchema
```
this doesn't exist anymore right?

thecacho replied:
The `cn.lua` does not exist anymore. Types however, are packed within that 'alias'.

ellisonleao replied:
> Types however, are packed within that 'alias'.

not sure if I follow. I know that the `cn` namespace is being used for other types, but I don't see this one in particular. Could you be more precise?

ellisonleao commented on lua/carbon-now.lua:
```diff
@@ -33,8 +38,13 @@ local function query_param_encode(str)
+---helper function to encode and concatenate a [k,v] table
+---as query parameters. iEx: {a = b, c = d} --> a=b&c=d
+---@param values table
+---@return string # Concatanation of the enconded query parameters
```
```suggestion
---@return string # Concatenation of the encoded query parameters
```

ellisonleao commented on lua/carbon-now.lua:
```diff
@@ -91,37 +103,47 @@ local function get_open_command()
 
-local function create_snippet(opts)
+---@param opts {args: string, line1: integer, line2: integer}
+local function create_snapshot(opts)
```
any reason to change the function name?

thecacho replied:
It's a typo. Snippet is a relative small amount of code used to excerpt into a larger codebase (autocompletion). Meanwhile the Snapshot, is a photograph, especially one taken quickly or in a given moment,to keep track of state or quick share.

ellisonleao replied:
makes sense, but since english is not my native language, we usually use the word "snippet" for both cases.

ellisonleao commented:
hey @Domenic-MZS i will keep the work from here! huge thanks on the emmylua and types stuff
